lib/driver.js

```diff
@@ -1684,11 +1686,14 @@ class XCUITestDriver extends BaseDriver {
     const {install, skipUninstall} = await this.checkAutInstallationState();
     if (install) {
       if (this.isRealDevice()) {
+        const shouldEnforceUninstall = this.opts.enforceAppInstall !== false || this.opts.fullReset;
```

Comment:
I am still not quite convinced why we need this flag. I thought we always perform uninstall if skipUninstall is not true

Reply:
That is to keep current behavior when  `enforceAppInstall: false` and `fullReset: false`, that expect not clear the device local state. So, if the device already had offload app for example, then the new session request may want to get the error without removing the app once and attempting to install it again.


`skipUninstall` can be true when wasAppInstalled is false like this offload app, or not installed yet. 
```
    if (enforceAppInstall !== false || fullReset || !wasAppInstalled) {
      return {
        install: true,
        skipUninstall: !wasAppInstalled,
      };
    }
```

Then, if the session does not require clearing the local (`enforceAppInstall: false` or `fullReset: false`), the install failure won't call `driver.remove` and `driver.install` even if the error is the `MismatchedApplicationIdentifierEntitlement`. Then, the current implementation raises an error

Reply:
> Then, if the session does not require clearing the local (enforceAppInstall: false or fullReset: false), the install failure won't call driver.remove and driver.install even if the error is the MismatchedApplicationIdentifierEntitlement

I am not quite sure why driver.remove won't be called in such case. We always try to remove an app if install fails with the particular error though